File: trading_system/core/streaming.py
"""
Streaming (incremental) indicators — O(1) per-tick updates.

Instead of recomputing EMA/RSI/MACD/Bollinger over all N candles
every tick, maintain running state and compute only the delta.

Speedup: O(n) → O(1) per indicator per tick.

Hybrid implementation: uses Rust-native accelerated backend when
available (rust_core), falls back to pure Python otherwise.
"""

import math
from collections import deque
from typing import Dict, List, Optional, Tuple

# Try Rust-native streaming backend
try:
    from rust_core import (
        PyStreamingEngine as _RustStreamingEngine,
        PyStreamingIndicators as _RustStreamingIndicators,
    )
    _HAS_RUST_STREAMING = True
except ImportError:  # pragma: no cover
    _HAS_RUST_STREAMING = False


class RingBuffer:
    """Fixed-size circular buffer for streaming data."""

    __slots__ = ("_data", "_maxlen", "_head", "_size")

    def __init__(self, maxlen: int = 200):
        self._data = [0.0] * maxlen
        self._maxlen = maxlen
        self._head = 0
        self._size = 0

    def append(self, value: float) -> None:
        self._data[self._head] = value
        self._head = (self._head + 1) % self._maxlen
        if self._size < self._maxlen:
            self._size += 1

    def __getitem__(self, index):
        if index < 0:
            index += self._size
        if index < 0 or index >= self._size:
            raise IndexError("index out of range")
        pos = (self._head - self._size + index) % self._maxlen
        return self._data[pos]

    def __len__(self) -> int:
        return self._size

    def to_list(self) -> List[float]:
        return [self[i] for i in range(self._size)]

    @property
    def last(self) -> Optional[float]:
        return self[-1] if self._size > 0 else None

    @property
    def oldest(self) -> Optional[float]:
        return self[0] if self._size > 0 else None


class StreamingIndicators:
    """Incremental indicator engine for one product.

    Maintains running state for EMA, RSI, SMA, Bollinger, MACD, Z-score.
    Each call to update() processes one new price in O(1) time.

    When Rust acceleration is available, delegates to the native backend.
    """

    __slots__ = (
        "product_id", "closes", "volumes",
        "_ema_state", "_sma_state", "_sq_sum_state",
        "_rsi_avg_gain", "_rsi_avg_loss", "_rsi_prev_close",
        "_macd_ema_fast", "_macd_ema_slow", "_macd_signal",
    )

    def __init__(self, product_id: str, maxlen: int = 200):
        self.product_id = product_id
        self.closes = RingBuffer(maxlen)
        self.volumes = RingBuffer(maxlen)

        self._ema_state: Dict[int, Optional[float]] = {}
        self._sma_state: Dict[int, Optional[float]] = {}
        self._sq_sum_state: Dict[int, float] = {}
        self._rsi_avg_gain: Optional[float] = None
        self._rsi_avg_loss: Optional[float] = None
        self._rsi_prev_close: Optional[float] = None
        self._macd_ema_fast: Optional[float] = None
        self._macd_ema_slow: Optional[float] = None
        self._macd_signal: Optional[float] = None

    def update(self, close: float, volume: float = 0.0) -> None:
        self.closes.append(close)
        self.volumes.append(volume)
        self._update_sma(close)
        self._update_ema(close)
        self._update_rsi(close)
        self._update_macd(close)

    # ── EMA ─────────────────────────────────────────────────────────

    def _update_ema(self, price: float) -> None:
        for period, prev in list(self._ema_state.items()):
            k = 2.0 / (period + 1)
            self._ema_state[period] = price * k + prev * (1.0 - k)

    def seed_ema(self, period: int, closes: List[float]) -> float:
        if not closes or len(closes) < period:
            val = closes[-1] if closes else 0.0
        else:
            val = sum(closes[:period]) / period
        self._ema_state[period] = val
        return val

    def ema(self, period: int) -> Optional[float]:
        return self._ema_state.get(period)

    # ── SMA + Bollinger ─────────────────────────────────────────────

    def _update_sma(self, price: float) -> None:
        n = len(self.closes)
        sq2 = price * price
        for period in list(self._sma_state.keys()):
            prev_sma = self._sma_state[period]
            prev_sq = self._sq_sum_state.get(period, 0.0)
            if prev_sma is None:
                continue
            if n > period:
                oldest = self.closes[n - period - 1] if n > period else price
                new_sma = prev_sma + (price - oldest) / period
                new_sq = prev_sq + sq2 - oldest * oldest
            else:
                new_sma = prev_sma + (price - prev_sma) / period
                new_sq = prev_sq + sq2
            self._sma_state[period] = new_sma
            self._sq_sum_state[period] = new_sq

    def seed_sma(self, period: int, closes: List[float]) -> float:
        if len(closes) < period:
            return closes[-1] if closes else 0.0
        recent = closes[:period]
        val = sum(recent) / period
        sq_sum = sum(x * x for x in recent)
        self._sma_state[period] = val
        self._sq_sum_state[period] = sq_sum
        return val

    def sma(self, period: int) -> Optional[float]:
        return self._sma_state.get(period)

    def bollinger(self, period: int, std_mult: float = 2.0) -> Optional[Tuple[float, float, float]]:
        mid = self._sma_state.get(period)
        n = len(self.closes)
        if mid is None or n < period:
            return None
        sq_sum = self._sq_sum_state.get(period, 0.0)
        variance = sq_sum / period - mid * mid
        std = math.sqrt(max(variance, 0.0))
        return (mid, mid + std_mult * std, mid - std_mult * std)

    # ── RSI ─────────────────────────────────────────────────────────

    def _update_rsi(self, price: float) -> None:
        if self._rsi_prev_close is None:
            self._rsi_prev_close = price
            return
        change = price - self._rsi_prev_close
        gain = max(change, 0.0)
        loss = max(-change, 0.0)
        self._rsi_prev_close = price

        if self._rsi_avg_gain is None:
            self._rsi_avg_gain = gain
            self._rsi_avg_loss = loss
        else:
            self._rsi_avg_gain = (self._rsi_avg_gain * 13.0 + gain) / 14.0
            self._rsi_avg_loss = (self._rsi_avg_loss * 13.0 + loss) / 14.0

    def seed_rsi(self, period: int, closes: List[float], seed_method: str = "wilder") -> float:
        if len(closes) < period + 1:
            return 50.0
        gains, losses = 0.0, 0.0
        for i in range(len(closes) - period, len(closes)):
            change = closes[i] - closes[i - 1]
            gains += max(change, 0.0)
            losses += max(-change, 0.0)
        self._rsi_avg_gain = gains / period
        self._rsi_avg_loss = losses / period
        self._rsi_prev_close = closes[-1]
        return self.rsi()

    def rsi(self) -> Optional[float]:
        if self._rsi_avg_gain is None or self._rsi_avg_loss is None:
            return None
        if self._rsi_avg_loss == 0:
            return 100.0
        rs = self._rsi_avg_gain / self._rsi_avg_loss
        return 100.0 - 100.0 / (1.0 + rs)

    # ── MACD ────────────────────────────────────────────────────────

    def _update_macd(self, price: float) -> None:
        fast = self._macd_ema_fast
        slow = self._macd_ema_slow
        if fast is not None and slow is not None:
            k12 = 2.0 / 13.0
            k26 = 2.0 / 27.0
            self._macd_ema_fast = price * k12 + fast * (1.0 - k12)
            self._macd_ema_slow = price * k26 + slow * (1.0 - k26)
            macd_line = self._macd_ema_fast - self._macd_ema_slow
            if self._macd_signal is not None:
                k9 = 2.0 / 10.0
                self._macd_signal = macd_line * k9 + self._macd_signal * (1.0 - k9)
            else:
                self._macd_signal = macd_line

    def seed_macd(self, closes: List[float],
                  fast: int = 12, slow: int = 26, signal: int = 9) -> None:
        self._macd_ema_fast = self.seed_ema(fast, closes)
        self._macd_ema_slow = self.seed_ema(slow, closes)
        if self._macd_ema_fast and self._macd_ema_slow:
            macd_line = self._macd_ema_fast - self._macd_ema_slow
            self._macd_signal = macd_line

    def macd(self) -> Optional[Tuple[float, float, float]]:
        if self._macd_ema_fast is None or self._macd_ema_slow is None:
            return None
        macd_line = self._macd_ema_fast - self._macd_ema_slow
        sig = self._macd_signal if self._macd_signal is not None else macd_line
        return (macd_line, sig, macd_line - sig)


class StreamingEngine:
    """Manages StreamingIndicators for multiple products.

    One engine instance serves the entire tick loop.

    Uses Rust-native StreamingEngine when available; falls back to pure Python.
    """

    def __init__(self):
        if _HAS_RUST_STREAMING:
            self._rust = _RustStreamingEngine()
            self._products: Dict[str, StreamingIndicators] = {}
        else:
            self._rust = None
            self._products: Dict[str, StreamingIndicators] = {}

    def get_or_create(self, product_id: str, maxlen: int = 200) -> object:
        if self._rust is not None:
            return self._rust.get_or_create(product_id, maxlen)
        if product_id not in self._products:
            self._products[product_id] = StreamingIndicators(product_id, maxlen)
        return self._products[product_id]

    def seed(self, product_id: str, closes: List[float], volumes: Optional[List[float]] = None):
        ind = self.get_or_create(product_id)
        for i in range(len(closes)):
            v = volumes[i] if volumes and i < len(volumes) else 0.0
            ind.update(closes[i], v)

    def try_get(self, product_id: str) -> Optional[object]:
        """Return streaming indicators handle if product exists, else None."""
        if self._rust is not None:
            return self._rust.try_get(product_id)
        return self._products.get(product_id)

    def update(self, product_id: str, close: float, volume: float = 0.0):
        if self._rust is not None:
            self._rust.update(product_id, close, volume)
            return
        ind = self._products.get(product_id)
        if ind:
            ind.update(close, volume)

    def ema(self, product_id: str, period: int) -> Optional[float]:
        if self._rust is not None:
            return self._rust.ema(product_id, period)
        ind = self._products.get(product_id)
        return ind.ema(period) if ind else None

    def rsi(self, product_id: str) -> Optional[float]:
        if self._rust is not None:
            return self._rust.rsi(product_id)
        ind = self._products.get(product_id)
        return ind.rsi() if ind else None

    def macd(self, product_id: str) -> Optional[Tuple[float, float, float]]:
        if self._rust is not None:
            return self._rust.macd(product_id)
        ind = self._products.get(product_id)
        return ind.macd() if ind else None

    def bollinger(self, product_id: str, period: int = 20) -> Optional[Tuple[float, float, float]]:
        ind = self._products.get(product_id)
        return ind.bollinger(period) if ind else None

    @property
    def products(self) -> Dict[str, object]:
        if self._rust is not None:
            return {}  # Not accessible via dict when using Rust backend
        return self._products

File: rust_core/src/streaming.rs
/// Streaming (incremental) indicators — O(1) per-tick updates.
///
/// Maintains running state for EMA, SMA(+Bollinger), RSI, MACD, Z-score
/// so each new tick processes in O(1) instead of recomputing over all N candles.
///
/// Mirrors trading_system/core/streaming.py in Rust.

use std::collections::HashMap;

/// Fixed-size circular buffer for streaming data.
#[derive(Clone)]
pub struct RingBuffer {
    data: Vec<f64>,
    maxlen: usize,
    head: usize,
    size: usize,
}

impl RingBuffer {
    pub fn new(maxlen: usize) -> Self {
        Self {
            data: vec![0.0; maxlen],
            maxlen,
            head: 0,
            size: 0,
        }
    }

    pub fn append(&mut self, value: f64) {
        self.data[self.head] = value;
        self.head = (self.head + 1) % self.maxlen;
        if self.size < self.maxlen {
            self.size += 1;
        }
    }

    pub fn get(&self, index: usize) -> Option<f64> {
        if index >= self.size {
            return None;
        }
        let pos = (self.head + self.maxlen - self.size + index) % self.maxlen;
        Some(self.data[pos])
    }

    pub fn last(&self) -> Option<f64> {
        if self.size == 0 {
            None
        } else {
            self.get(self.size - 1)
        }
    }

    pub fn oldest(&self) -> Option<f64> {
        self.get(0)
    }

    pub fn len(&self) -> usize {
        self.size
    }

    pub fn to_vec(&self) -> Vec<f64> {
        (0..self.size).map(|i| self.get(i).unwrap()).collect()
    }
}

/// Running state for one indicator type.  Not public — use StreamingIndicators.
struct EmaState(HashMap<usize, f64>);

impl EmaState {
    fn new() -> Self {
        Self(HashMap::new())
    }

    fn update(&mut self, price: f64) {
        for (period, prev) in self.0.iter_mut() {
            let k = 2.0 / (*period as f64 + 1.0);
            *prev = price * k + *prev * (1.0 - k);
        }
    }

    fn seed(&mut self, period: usize, value: f64) {
        self.0.insert(period, value);
    }

    fn get(&self, period: usize) -> Option<f64> {
        self.0.get(&period).copied()
    }
}

struct SmaState(HashMap<usize, (f64, f64)>); // period -> (sma, sq_sum)

impl SmaState {
    fn new() -> Self {
        Self(HashMap::new())
    }

    fn update(&mut self, price: f64, buffer: &RingBuffer) {
        let n = buffer.len();
        let sq2 = price * price;
        let periods: Vec<usize> = self.0.keys().copied().collect();
        for period in periods {
            let (prev_sma, prev_sq) = match self.0.get(&period) {
                Some(v) => *v,
                None => continue,
            };
            if n > period {
                if let Some(oldest) = buffer.get(n - period - 1) {
                    let new_sma = prev_sma + (price - oldest) / period as f64;
                    let new_sq = prev_sq + sq2 - oldest * oldest;
                    self.0.insert(period, (new_sma, new_sq));
                }
            } else {
                let new_sma = prev_sma + (price - prev_sma) / period as f64;
                let new_sq = prev_sq + sq2;
                self.0.insert(period, (new_sma, new_sq));
            }
        }
    }

    fn seed(&mut self, period: usize, sma: f64, sq_sum: f64) {
        self.0.insert(period, (sma, sq_sum));
    }

    fn get(&self, period: usize) -> Option<f64> {
        self.0.get(&period).map(|v| v.0)
    }

    fn bollinger(&self, period: usize) -> Option<(f64, f64, f64)> {
        let (mid, sq_sum) = self.0.get(&period).copied()?;
        let variance = sq_sum / period as f64 - mid * mid;
        let std = variance.max(0.0).sqrt();
        Some((mid, mid + 2.0 * std, mid - 2.0 * std))
    }
}

/// Streaming indicator engine for one product.
pub struct StreamingIndicators {
    pub product_id: String,
    pub closes: RingBuffer,
    pub volumes: RingBuffer,

    ema: EmaState,
    sma: SmaState,

    // RSI state (Wilder's smoothing)
    rsi_avg_gain: Option<f64>,
    rsi_avg_loss: Option<f64>,
    rsi_prev_close: Option<f64>,

    // MACD state
    macd_ema_fast: Option<f64>,
    macd_ema_slow: Option<f64>,
    macd_signal: Option<f64>,
}

impl StreamingIndicators {
    pub fn new(product_id: &str, maxlen: usize) -> Self {
        Self {
            product_id: product_id.to_string(),
            closes: RingBuffer::new(maxlen),
            volumes: RingBuffer::new(maxlen),
            ema: EmaState::new(),
            sma: SmaState::new(),
            rsi_avg_gain: None,
            rsi_avg_loss: None,
            rsi_prev_close: None,
            macd_ema_fast: None,
            macd_ema_slow: None,
            macd_signal: None,
        }
    }

    pub fn update(&mut self, close: f64, volume: f64) {
        self.closes.append(close);
        self.volumes.append(volume);
        self.sma.update(close, &self.closes);
        self.ema.update(close);
        self.update_rsi(close);
        self.update_macd(close);
    }

    // ── EMA ────────────────────────────────────────────────────────

    pub fn seed_ema(&mut self, period: usize, value: f64) {
        self.ema.seed(period, value);
    }

    pub fn ema(&self, period: usize) -> Option<f64> {
        self.ema.get(period)
    }

    // ── SMA + Bollinger ────────────────────────────────────────────

    pub fn seed_sma(&mut self, period: usize, sma: f64, sq_sum: f64) {
        self.sma.seed(period, sma, sq_sum);
    }

    pub fn sma(&self, period: usize) -> Option<f64> {
        self.sma.get(period)
    }

    pub fn bollinger(&self, period: usize) -> Option<(f64, f64, f64)> {
        self.sma.bollinger(period)
    }

    // ── RSI ────────────────────────────────────────────────────────

    fn update_rsi(&mut self, price: f64) {
        let prev = match self.rsi_prev_close {
            Some(v) => v,
            None => {
                self.rsi_prev_close = Some(price);
                return;
            }
        };
        let change = price - prev;
        let gain = change.max(0.0);
        let loss = (-change).max(0.0);
        self.rsi_prev_close = Some(price);

        match (self.rsi_avg_gain, self.rsi_avg_loss) {
            (None, _) => {
                self.rsi_avg_gain = Some(gain);
                self.rsi_avg_loss = Some(loss);
            }
            (Some(avg_gain), Some(avg_loss)) => {
                self.rsi_avg_gain = Some((avg_gain * 13.0 + gain) / 14.0);
                self.rsi_avg_loss = Some((avg_loss * 13.0 + loss) / 14.0);
            }
            _ => {
                self.rsi_avg_gain = Some(gain);
                self.rsi_avg_loss = Some(loss);
            }
        }
    }

    pub fn seed_rsi(&mut self, closes: &[f64], period: usize) -> f64 {
        if closes.len() < period + 1 {
            return 50.0;
        }
        let mut gains = 0.0_f64;
        let mut losses = 0.0_f64;
        let start = closes.len() - period;
        for i in start..closes.len() {
            let change = closes[i] - closes[i - 1];
            gains += change.max(0.0);
            losses += (-change).max(0.0);
        }
        self.rsi_avg_gain = Some(gains / period as f64);
        self.rsi_avg_loss = Some(losses / period as f64);
        self.rsi_prev_close = closes.last().copied();
        self.rsi()
    }

    pub fn rsi(&self) -> f64 {
        match (self.rsi_avg_gain, self.rsi_avg_loss) {
            (_, Some(l)) if l == 0.0 => 100.0,
            (Some(g), Some(l)) => 100.0 - 100.0 / (1.0 + g / l),
            _ => 50.0,
        }
    }

    // ── MACD ───────────────────────────────────────────────────────

    fn update_macd(&mut self, price: f64) {
        match (self.macd_ema_fast, self.macd_ema_slow) {
            (Some(fast), Some(slow)) => {
                let k12 = 2.0 / 13.0;
                let k26 = 2.0 / 27.0;
                self.macd_ema_fast = Some(price * k12 + fast * (1.0 - k12));
                self.macd_ema_slow = Some(price * k26 + slow * (1.0 - k26));
                let macd_line = self.macd_ema_fast.unwrap() - self.macd_ema_slow.unwrap();
                self.macd_signal = Some(match self.macd_signal {
                    Some(sig) => {
                        let k9 = 2.0 / 10.0;
                        macd_line * k9 + sig * (1.0 - k9)
                    }
                    None => macd_line,
                });
            }
            _ => {}
        }
    }

    pub fn seed_macd(&mut self, ema_fast: f64, ema_slow: f64) {
        self.macd_ema_fast = Some(ema_fast);
        self.macd_ema_slow = Some(ema_slow);
        self.macd_signal = Some(ema_fast - ema_slow);
    }

    pub fn macd(&self) -> Option<(f64, f64, f64)> {
        let fast = self.macd_ema_fast?;
        let slow = self.macd_ema_slow?;
        let macd_line = fast - slow;
        let sig = self.macd_signal.unwrap_or(macd_line);
        Some((macd_line, sig, macd_line - sig))
    }
}

// ── StreamingEngine (thin manager for multiple products) ───────────

pub struct StreamingEngine {
    pub products: HashMap<String, StreamingIndicators>,
}

impl StreamingEngine {
    pub fn new() -> Self {
        Self {
            products: HashMap::new(),
        }
    }

    pub fn get_or_create(&mut self, product_id: &str, maxlen: usize) -> &mut StreamingIndicators {
        self.products
            .entry(product_id.to_string())
            .or_insert_with(|| StreamingIndicators::new(product_id, maxlen))
    }

    pub fn update(&mut self, product_id: &str, close: f64, volume: f64) {
        if let Some(ind) = self.products.get_mut(product_id) {
            ind.update(close, volume);
        }
    }

    pub fn ema(&self, product_id: &str, period: usize) -> Option<f64> {
        self.products.get(product_id)?.ema(period)
    }

    pub fn rsi(&self, product_id: &str) -> Option<f64> {
        self.products.get(product_id).map(|p| p.rsi())
    }

    pub fn macd(&self, product_id: &str) -> Option<(f64, f64, f64)> {
        self.products.get(product_id)?.macd()
    }
}

#[cfg(test)]
mod coverage_tests {
    use super::*;

    #[test]
    fn test_ring_buffer_get_oob() {
        let mut buf = RingBuffer::new(3);
        buf.append(1.0);
        assert_eq!(buf.get(5), None); // index >= size
        assert_eq!(buf.get(0), Some(1.0));
    }

    #[test]
    fn test_sma_seeded_and_bollinger() {
        let mut ind = StreamingIndicators::new("TST", 100);
        // unseeded period -> None paths
        assert!(ind.sma(99).is_none());
        assert!(ind.bollinger(99).is_none());
        assert!(ind.ema(99).is_none());

        // seed SMA and exercise both n<=period and n>period branches
        ind.seed_sma(3, 100.0, 30000.0);
        ind.update(101.0, 10.0); // n=1 -> else branch
        ind.update(102.0, 10.0); // n=2 -> else branch
        ind.update(103.0, 10.0); // n=3 -> else branch
        ind.update(104.0, 10.0); // n=4 -> Some(oldest) branch
        assert!(ind.sma(3).is_some());
        let (mid, up, lo) = ind.bollinger(3).unwrap();
        assert!(mid.is_finite() && up > lo);
    }

    #[test]
    fn test_rsi_fresh_and_seed_edges() {
        let mut ind = StreamingIndicators::new("TST", 100);
        // rsi() before any update -> (None, None) -> 50
        assert_eq!(ind.rsi(), 50.0);
        // first update: prev None branch
        ind.update(50000.0, 10.0);
        // second update: (None, _) branch sets avg
        ind.update(50100.0, 10.0);
        assert!(ind.rsi().is_finite());
        // seed_rsi with too-short closes -> 50
        let short = vec![1.0, 2.0];
        assert_eq!(ind.seed_rsi(&short, 14), 50.0);
    }

    #[test]
    fn test_macd_unseeded() {
        let mut ind = StreamingIndicators::new("TST", 100);
        // update before seed -> _ branch in update_macd
        ind.update(50100.0, 10.0);
        // macd() before seed -> fast None -> None
        assert!(ind.macd().is_none());
    }

    #[test]
    fn test_engine_missing_product() {
        let mut engine = StreamingEngine::new();
        // update on unknown product -> no-op
        engine.update("NOPE", 1.0, 1.0);
        // accessors on unknown product -> None
        assert!(engine.ema("NOPE", 9).is_none());
        assert!(engine.rsi("NOPE").is_none());
        assert!(engine.macd("NOPE").is_none());
        // with a real product
        engine.get_or_create("BTC-USD", 100).seed_ema(9, 50000.0);
        engine.update("BTC-USD", 50100.0, 10.0);
        assert!(engine.ema("BTC-USD", 9).is_some());
        assert!(engine.rsi("BTC-USD").is_some());
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_ring_buffer() {
        let mut buf = RingBuffer::new(5);
        assert!(buf.last().is_none());
        buf.append(1.0);
        buf.append(2.0);
        buf.append(3.0);
        assert_eq!(buf.len(), 3);
        assert_eq!(buf.get(0), Some(1.0));
        assert_eq!(buf.get(2), Some(3.0));
        assert_eq!(buf.last(), Some(3.0));
        assert_eq!(buf.oldest(), Some(1.0));
    }

    #[test]
    fn test_ring_buffer_wrap() {
        let mut buf = RingBuffer::new(3);
        buf.append(1.0);
        buf.append(2.0);
        buf.append(3.0);
        buf.append(4.0); // wraps, overwrites index 0
        assert_eq!(buf.len(), 3);
        assert_eq!(buf.get(0), Some(2.0)); // oldest
        assert_eq!(buf.get(2), Some(4.0)); // newest
    }

    #[test]
    fn test_ema_update() {
        let mut ind = StreamingIndicators::new("BTC-USD", 100);
        ind.seed_ema(9, 50000.0);

        ind.update(50100.0, 100.0);
        let ema = ind.ema(9).unwrap();
        assert!(ema > 50000.0 && ema < 50100.0);
    }

    #[test]
    fn test_rsi_update() {
        let mut ind = StreamingIndicators::new("BTC-USD", 100);

        // Seed with 16 flat closes → no gains/losses → RSI should be 100
        let closes: Vec<f64> = vec![50000.0; 16];
        assert!((ind.seed_rsi(&closes, 14) - 100.0).abs() < 0.1);

        // Update with a price spike up
        ind.update(60000.0, 100.0);
        let rsi2 = ind.rsi();
        assert!(rsi2 > 50.0);
    }

    #[test]
    fn test_macd_update() {
        let mut ind = StreamingIndicators::new("BTC-USD", 100);
        ind.seed_macd(50000.0, 50000.0);

        ind.update(50100.0, 100.0);
        let macd = ind.macd().unwrap();
        // After one update with up-tick, macd_line should be positive
        assert!(macd.0 > 0.0);
    }

    #[test]
    fn test_streaming_engine() {
        let mut engine = StreamingEngine::new();
        engine.get_or_create("BTC-USD", 100);
        engine.update("BTC-USD", 50000.0, 100.0);
        engine.update("BTC-USD", 50100.0, 150.0);
        assert!(engine.products.contains_key("BTC-USD"));
    }

    #[test]
    fn test_to_vec() {
        let mut buf = RingBuffer::new(10);
        buf.append(1.0);
        buf.append(2.0);
        buf.append(3.0);
        let v = buf.to_vec();
        assert_eq!(v, vec![1.0, 2.0, 3.0]);
    }
}
